dango/account-margin/src/query.rs

```diff
@@ -1,16 +1,90 @@
 use {
+    anyhow::bail,
     dango_auth::NEXT_SEQUENCE,
-    dango_types::account::single::QueryMsg,
-    grug::{ImmutableCtx, Json, JsonSerExt, StdResult, Storage},
+    dango_lending::DEBTS,
+    dango_types::{
+        account::margin::{HealthResponse, QueryMsg},
+        config::LENDING_KEY,
+        lending::{CollateralPower, LendingAppConfig},
+    },
+    grug::{
+        Addr, BorshDeExt, Coins, Denom, ImmutableCtx, Inner, IsZero, Json, JsonSerExt, NumberConst,
+        QuerierWrapper, StdResult, Storage, Udec128,
+    },
+    std::collections::BTreeMap,
 };
 
 #[cfg_attr(not(feature = "library"), grug::export)]
-pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> StdResult<Json> {
+pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> anyhow::Result<Json> {
     match msg {
         QueryMsg::Sequence {} => query_sequence(ctx.storage)?.to_json_value(),
+        QueryMsg::Health {} => query_health(ctx)?.to_json_value(),
     }
+    .map_err(Into::into)
 }
 
 fn query_sequence(storage: &dyn Storage) -> StdResult<u32> {
     NEXT_SEQUENCE.current(storage)
 }
+
+/// Calculates the health of a margin account.
+pub fn calculate_account_health(
+    querier: &QuerierWrapper,
+    margin_account: Addr,
+    debts: Coins,
+    collateral_powers: BTreeMap<Denom, CollateralPower>,
+) -> anyhow::Result<HealthResponse> {
+    // Calculate the total value of the debts.
+    let mut total_debt_value = Udec128::ZERO;
+    for coin in debts {
```

Comment:
The function `dango_oracle::raw_query_price(querier, &coin.denom)?` query the oracle contract. If the user has different debts is a waste. Add ad an optional param `maybe_oracle: Option<Addr>`. If it's None, query for it, otherwise use the one passed